src/mono/mono/mini/interp/interpgo.c

```diff
@@ -0,0 +1,395 @@
+	//  load a new interpgo table during the search and then attempt to g_free the old table's
+	//  data out from under us, and we would crash. This means that multiple threads can't
+	//  perform a search in this table at once. It would be nice if we had a rwlock for this
+	// In disable_threads builds and situations where no table is loaded this doesn't matter
+	// Another solution would be to make the buffer ref-counted, so that it only gets freed
+	//  once nobody is using it. But that's worse in the no concurrent access case since
+	//  we'd waste cycles modifying the refcount when it doesn't matter
+	mono_os_mutex_lock (&table->mutex);
+	void * result = mono_binary_search (hash, table->data, table->size / MM3_HASH_BYTE_SIZE, MM3_HASH_BYTE_SIZE, hash_comparer);
+	mono_os_mutex_unlock (&table->mutex);
+	return (result != NULL);
+}
+
+static void
+table_add (interpgo_table **table_variable, uint8_t hash[MM3_HASH_BYTE_SIZE]) {
+	interpgo_table *table = *table_variable, *temp;
+	// If we don't have a table yet, allocate one
+	if (!table) {
+		table = g_malloc0 (sizeof (interpgo_table));
+		// We may have lost a race against another thread that just allocated a table
+		temp = mono_atomic_cas_ptr ((volatile gpointer *)table_variable, table, NULL);
+		if (temp != NULL) {
+			// We lost a race, so destroy our new table and use the one that someone else allocated
+			g_free (table);
+			table = (interpgo_table *)temp;
+		}
+	}
+
+	// Now we have a pointer to the one and only table associated with table_variable. Lock it.
+	mono_os_mutex_lock (&table->mutex);
+
+	const uint32_t required_size = table->size + MM3_HASH_BYTE_SIZE,
+		required_capacity = MAX (required_size, TABLE_MINIMUM_SIZE);
+
+	// If we're out of space or haven't yet allocated a buffer for this table, calculate
+	//  an appropriate larger size and grow/allocate the buffer. We start at a fixed size,
+	//  then after that grow the current size by a set ratio per step.
+	while (required_capacity >= table->capacity) {
+		uint32_t new_capacity = MAX (required_capacity, (table->capacity * TABLE_GROWTH_FACTOR / 100));
+		if (table->data)
+			table->data = g_realloc (table->data, new_capacity);
+		else
+			table->data = g_malloc0 (new_capacity);
+		table->capacity = new_capacity;
+	}
+
+	// Copy the whole hash into the table at the end and update the size of the data
+	memcpy (table->data + table->size, hash, MM3_HASH_BYTE_SIZE);
+	table->size = required_size;
+
+	mono_os_mutex_unlock (&table->mutex);
+}
+
+static void
+table_sort_locked (interpgo_table *table) {
+	mono_qsort (table->data, table->size / MM3_HASH_BYTE_SIZE, MM3_HASH_BYTE_SIZE, hash_comparer);
+}
+
+static void
+compute_method_hash (MonoMethod *method, uint8_t outbuf[MM3_HASH_BYTE_SIZE]) {
+	// method token + image guid
+	size_t size = sizeof(uint32_t) + 16;
+	uint32_t *inbuf = alloca (size);
+	// method tokens are globally unique within a given assembly
+	inbuf[0] = mono_method_get_token (method);
```

Comment:
This is not true for generic instances.

Reply:
OK, so I need to hash the generic arguments? Or is it more complex than that?

Reply:
or skip generic instances ? Are they usually hot ?

Reply:
some generic instances are very hot in the BCL, like the ones used in string/array searches

Reply:
It can hash the full method name for example, but maybe its good enough this way.

Reply:
If it's a wrapper method, I think the token is either shared with the method that it is wrapping, or it is zero.  maybe we don't mind the collisions?

Reply:
Can we share the AOT compiler's logic for uniquely identifying methods?

Reply:
The aot compiler/runtime has a mono_aot_method_hash () which returns a good hash.

Reply:
mono_aot_method_hash looks comprehensive but also looks really expensive, and i'm not sure how well it will avoid collisions. do you still want me to use it?